controller/canary/traffic_reconciler.go

```
@@ -0,0 +1,455 @@
+		}
+		logger = logger.WithField("baseline", d.baseline.Name)
+
+		totalActualTraffic := totalActualTrafficFor([]canaryStack{d.green, d.blue, *d.baseline})
+		if totalActualTraffic.Error() != nil {
+			logger.Errorf(
+				"failed to compute green+blue+baseline traffic: %s", totalActualTraffic.Error())
+			continue
+		}
+
+		state := tsState{
+			greenActualTraffic:    d.green.actualTraffic,
+			blueName:              d.blue.Name,
+			blueActualTraffic:     d.blue.actualTraffic,
+			baselineName:          d.baseline.Name,
+			baselineActualTraffic: d.baseline.actualTraffic,
+			desiredSwitchDuration: d.desiredSwitchDuration,
+			switchStartTime:       d.switchStartTime,
+			totalActualTraffic:    totalActualTraffic,
+			ingressAction:         actionDoNothing,
+		}
+		e := r.env.WithLogger(logger)
+		if d.switchStartTime.IsZero() || d.green.actualTraffic.IsZero() {
+			if !manageStateForNoTraffic(e, &state, *d, &ssc.StackSet) {
+				continue
+			}
+		} else if r.checker.ExpectationsFulfilled(*d.green.Stack, *d.baseline.Stack) {
+			logger.Infof("expectations still fulfilled, continuing")
+			ok := manageStateForExpectationsFulfilled(e, &state, *d, &ssc.StackSet, totalActualTraffic)
+			if !ok {
+				continue
+			}
+		} else {
+			state.ingressAction = actionRollback
+			r.env.Client.MarkStackNotGreen(&ssc.StackSet, d.green.Stack, "rollback")
+		}
+		newGreenStates[d.green.Name] = state
+	}
+	r.greenStates = newGreenStates // Also cleans up the old states.
+	return nil
+}
+
+// tsInput is a helper type for ReconcileStacks and its own helper functions. Please do not use it outside.
+type tsInput struct {
+	green                 canaryStack
+	blue                  canaryStack
+	baseline              *canaryStack
+	desiredSwitchDuration time.Duration
+	switchStartTime       time.Time
+}
+
+func newTsInput(green *zv1.Stack, allStacks map[types.UID]*entities.StackContainer, tm TrafficMap) (*tsInput, []error) {
```

What's the idea behind returning a slice of errors here? Why not just return the first error to fail fast?

---

This to collectively tell the user whats wrong. This was done with the idea that since this is pretty quick in just reading annotations, all the errors are sent in one go thereby saving the user lots of round trips of redeploying.